azure/services/managedclusters/managedclusters.go

```diff
@@ -105,17 +108,22 @@ func postCreateOrUpdateResourceHook(ctx context.Context, scope ManagedClusterSco
 		})
 	}
 
+	if scope.IsManagedVersionUpgrade() && managedCluster.Spec.KubernetesVersion != nil {
+		kubernetesVersion := fmt.Sprintf("v%s", *managedCluster.Spec.KubernetesVersion)
```

Comment:
When auto upgrades are in place, we do not propagate the new version to the spec of the ASO resource. We should be looking this in the status.

Reply:
Should we update this then?

Reply:
@nojnhuh good question. 

I think this PR has an issue and i am a bit confused about how to solve it. 
This PR was good prior to ASO getting merged and when CAPZ was making an api call to create the AKS resource.

When auto upgrade is set AKS manages the version. Upgrades ideally have to be completely managed by AKS.

But AKS also provides the users options to upgrade AKS to higher version than the existing one if required. 

So now there are 2 sources which had the control for updating AKS version.
1.) AKS itself based on the upgrade channel
2.) The user if he wants the version to be greater than the version chosen based on the upgrade channel. 

So what i did to handle both the cases is, to pick the higher version out of the 2 i.e version managed by AKS vs version managed by user. 

I was able to get the version managed by AKS from the object returned by the api call and always set it to the autupgradeversion in status. 

Now the problem is I don't know the version set by AKS. Or in other words how do i get the version managed by AKS ??

Reply:
Also I don't know if allowing users to update a version after setting an upgrade channel is a bug in AKS itself. 

Because that kinda conflicts with the whole idea of having an upgrade channel.

Reply:
> Also I don't know if allowing users to update a version after setting an upgrade channel is a bug in AKS itself.
> 
> Because that kinda conflicts with the whole idea of having an upgrade channel.

I could be convinced that that should be allowed to enable users to have a more conservative auto upgrade policy while allowing larger upgrades to be done manually. If you find docs that imply users shouldn't touch the version at all when auto upgrades are enabled then that seems worth bringing up to AKS.

Reply:
> In this case though it makes sense to me to populate the AzureManagedControlPlane's status based on the ASO ManagedCluster's status.

Right will make that update and test it, because with the existing code after i test it does not work as expected.

> If you find docs that imply users shouldn't touch the version at all when auto upgrades are enabled then that seems worth bringing up to AKS.

I don't see the anywhere in the docs that say we shouldn't be touching the version when an auto upgrade is configured for the AKS cluster. It was just a thought in my head, that we are doing two conflicting things at the same time. Like on 1 hand we say auto upgrade and on the other hand we say i would upgrade.

Reply:
@nojnhuh Created a cluster with 1.26.6 k8s version and upgrade channel was set to stable. The cluster got auto upgraded to 1.27.7. Digged through the logs of ASO 

We see the logs do print the version as 1.27.7

`I0115 19:17:09.559965       1 azure_generic_arm_reconciler_instance.go:579] "msg"="Got ARM status" "azureName"="rnl-capz-aks-auto-upgrade" "logger"="controllers.ManagedClusterController" "name"="rnl-capz-aks-auto-upgrade" "namespace"="default" "status"="{\"id\":\"/subscriptions/<<<<subscription-id>>>>/resourcegroups/lochan-rg/providers/Microsoft.ContainerService/managedClusters/rnl-capz-aks-auto-upgrade\",\"identity\":{\"delegatedResources\":null,\"principalId\":\"<<<<principal-id>>>>\",\"tenantId\":\"<<<<tenant-id>>>>\",\"type\":\"SystemAssigned\",\"userAssignedIdentities\":null},\"location\":\"eastus\",\"name\":\"rnl-capz-aks-auto-upgrade\",\"properties\":{\"addonProfiles\":null,\"agentPoolProfiles\":[{\"availabilityZones\":null,\"count\":3,\"currentOrchestratorVersion\":\"1.27.7\",\"enableAutoScaling\":false,\"enableFIPS\":false,\"kubeletDiskType\":\"OS\",\"maxPods\":30,\"mode\":\"User\",\"name\":\"pool1\",\"nodeImageVersion\":\"AKSUbuntu-2204gen2containerd-202312.06.0\",\"nodeLabels\":null,\"nodeTaints\":null,\"orchestratorVersion\":\"1.27.7\",\"osDiskSizeGB\":128,\"osDiskType\":\"Managed\",\"osSKU\":\"Ubuntu\",\"osType\":\"Linux\",\"powerState\":{\"code\":\"Running\"},\"provisioningState\":\"Succeeded\",\"scaleDownMode\":\"Delete\",\"tags\":null,\"type\":\"VirtualMachineScaleSets\",\"vmSize\":\"Standard_D2s_v3\",\"vnetSubnetID\":\"/subscriptions/<<<<subscription-id>>>>/resourceGroups/lochan-rg/providers/Microsoft.Network/virtualNetworks/rnl-capz-aks-auto-upgrade/subnets/rnl-capz-aks-auto-upgrade\"},{\"availabilityZones\":null,\"count\":3,\"currentOrchestratorVersion\":\"1.27.7\",\"enableAutoScaling\":false,\"enableFIPS\":false,\"kubeletDiskType\":\"OS\",\"maxPods\":30,\"mode\":\"System\",\"name\":\"pool0\",\"nodeImageVersion\":\"AKSUbuntu-2204gen2containerd-202312.06.0\",\"nodeLabels\":null,\"nodeTaints\":null,\"orchestratorVersion\":\"1.27.7\",\"osDiskSizeGB\":128,\"osDiskType\":\"Managed\",\"osSKU\":\"Ubuntu\",\"osType\":\"Linux\",\"powerState\":{\"code\":\"Running\"},\"provisioningState\":\"Succeeded\",\"scaleDownMode\":\"Delete\",\"tags\":null,\"type\":\"VirtualMachineScaleSets\",\"vmSize\":\"Standard_D2s_v3\",\"vnetSubnetID\":\"/subscriptions/<<<<subscription-id>>>>/resourceGroups/lochan-rg/providers/Microsoft.Network/virtualNetworks/rnl-capz-aks-auto-upgrade/subnets/rnl-capz-aks-auto-upgrade\"}],\"autoUpgradeProfile\":{\"nodeOSUpgradeChannel\":\"NodeImage\",\"upgradeChannel\":\"stable\"},\"azurePortalFQDN\":\"rnl-capz-aks-auto-upgrade-n55o6664.portal.hcp.eastus.azmk8s.io\",\"currentKubernetesVersion\":\"1.27.7\",\"dnsPrefix\":\"rnl-capz-aks-auto-upgrade\",\"enableRBAC\":true,\"fqdn\":\"rnl-capz-aks-auto-upgrade-n55o6664.hcp.eastus.azmk8s.io\",\"identityProfile\":{\"kubeletidentity\":{\"clientId\":\"<<<<client-id>>>>\",\"objectId\":\"<<<<object-id>>>>\",\"resourceId\":\"/subscriptions/<<<<subscription-id>>>>/resourcegroups/MC_lochan-rg_rnl-capz-aks-auto-upgrade_eastus/providers/Microsoft.ManagedIdentity/userAssignedIdentities/rnl-capz-aks-auto-upgrade-agentpool\"}},\"kubernetesVersion\":\"1.27.7\",\"linuxProfile\":{\"adminUsername\":\"azureuser\",\"ssh\":{\"publicKeys\":[{\"keyData\":\"<ssh-key>\"}]}},\"maxAgentPools\":100,\"networkProfile\":{\"dnsServiceIP\":\"192.168.0.10\",\"ipFamilies\":[\"IPv4\"],\"loadBalancerProfile\":{\"effectiveOutboundIPs\":[{\"id\":\"/subscriptions/<<<<subscription-id>>>>/resourceGroups/MC_lochan-rg_rnl-capz-aks-auto-upgrade_eastus/providers/Microsoft.Network/publicIPAddresses/<<<<publicIPId>>>>\"}],\"managedOutboundIPs\":{\"count\":1}},\"loadBalancerSku\":\"Standard\",\"networkDataplane\":\"azure\",\"networkPlugin\":\"azure\",\"outboundType\":\"loadBalancer\",\"podCidrs\":null,\"serviceCidr\":\"192.168.0.0/16\",\"serviceCidrs\":[\"192.168.0.0/16\"]},\"nodeResourceGroup\":\"MC_lochan-rg_rnl-capz-aks-auto-upgrade_eastus\",\"oidcIssuerProfile\":{\"enabled\":false},\"powerState\":{\"code\":\"Running\"},\"privateLinkResources\":null,\"provisioningState\":\"Succeeded\",\"resourceUID\":\"<<<<resouce-id>>>>\",\"securityProfile\":{},\"servicePrincipalProfile\":{\"clientId\":\"msi\"},\"storageProfile\":{\"diskCSIDriver\":{\"enabled\":true},\"fileCSIDriver\":{\"enabled\":true},\"snapshotController\":{\"enabled\":true}},\"supportPlan\":\"KubernetesOfficial\",\"windowsProfile\":{\"adminUsername\":\"azureuser\",\"enableCSIProxy\":true},\"workloadAutoScalerProfile\":{}},\"sku\":{\"name\":\"Base\",\"tier\":\"Free\"},\"tags\":{\"Name\":\"rnl-capz-aks-auto-upgrade\",\"sigs.k8s.io_cluster-api-provider-azure_cluster_rnl-capz-aks-auto-upgrade\":\"owned\",\"sigs.k8s.io_cluster-api-provider-azure_role\":\"common\"},\"type\":\"Microsoft.ContainerService/ManagedClusters\"}"`



But the ManagedCluster object is not getting set/updated with version 1.27.7 it is still at 1.26.6

Reply:
Looks like some bug in the ASO.

Reply:
@LochanRn Hm, that's not I would expect either. Might be best to follow up in an ASO issue (https://github.com/azure/azure-service-operator) or on k8s Slack (#azure-service-operator).

Reply:
I started a slack thread: https://kubernetes.slack.com/archives/C046DEVLAQM/p1705423946331469